src/ol/render/canvas.js

```diff
@@ -437,7 +437,11 @@ export function drawImageOrLabel(
   context.save();
 
   if (opacity !== 1) {
-    context.globalAlpha *= opacity;
+    if (context.globalAlpha === undefined) {
+      context.globalAlpha = (context) => (context.globalAlpha *= opacity);
```

Comment:
A `ZIndexContextProxy` (i.e. what's returned by `ZIndexContext#getContext()`) will return `undefined` for `context.globalAlpha`. This is because `globalAlpha` is a state which is only maintained by the final rendering context, but not by the proxy, which only contains instructions, not state.

To set a `context.globalAlpha` that depends on a previous value thereof, we need to set a function on the `ZIndexContextProxy`, which is called with the target context as argument, and returns the desired value of the `globalAlpha` property.